src/url.rs

```diff
@@ -37,6 +40,86 @@ impl fmt::Display for Scheme {
     }
 }
 
+#[derive(PartialEq, Debug, Clone)]
+pub struct HttpUrl {
+    url: Url,
+    scheme: Scheme,
+    host: String,
+}
+
+impl HttpUrl {
+    pub fn port(&self) -> Result<u16> {
+        match self.url.port_or_known_default() {
+            Some(v) => Ok(v),
+            None => Err(Error::UrlError(format!(
+                "unsupported URL scheme {}: no port and no known default",
+                self.url.scheme()
+            ))),
+        }
+    }
+    pub fn scheme(&self) -> Scheme {
+        self.scheme.clone()
+    }
+    pub fn host(&self) -> &str {
+        &self.host
+    }
+    // give use an option to use the origin url struct and save us from porting too many codes
+    pub fn url(&self) -> &Url {
+        &self.url
+    }
+}
+
+impl TryFrom<Url> for HttpUrl {
+    type Error = Error;
+
+    fn try_from(url: Url) -> Result<Self> {
+        use core::str::FromStr;
+
+        let scheme = Scheme::from_str(url.scheme())?;
+        if scheme.ne(&Scheme::Http) && scheme.ne(&Scheme::Https) {
+            return Err(Error::UrlError(format!(
+                "unsupported URL scheme {}",
+                url.scheme()
+            )));
+        };
+        let Some(host) = url.host_str() else {
+            // people can use url.set_host(Option::None) to make host invalid, and currently we
```

Comment:
that is a good point to make, but also I think some URLs just don't have a host from the get-go, so that sounds like the two situations this is covering and as you have the comment it feels like you should mention the other one

Reply:
after testing, I think the url crate have ensure that there must have host in http and https.
I will change it to just unwrap, and add a test for declaration